<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Multiscale bootstrap cluster analysis. — clusterBoot • OpenRepGrid</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multiscale bootstrap cluster analysis. — clusterBoot"><meta name="description" content="p-values are calculated for each branch of the cluster dendrogram to indicate the stability of a specific partition.
clusterBoot will yield the same clusters as the cluster() function (i.e. standard hierarchical clustering) with
additional p-values. Two kinds of p-values are reported: bootstrap probabilities (BP) and approximately unbiased
(AU) probabilities (see Details section for more information)."><meta property="og:description" content="p-values are calculated for each branch of the cluster dendrogram to indicate the stability of a specific partition.
clusterBoot will yield the same clusters as the cluster() function (i.e. standard hierarchical clustering) with
additional p-values. Two kinds of p-values are reported: bootstrap probabilities (BP) and approximately unbiased
(AU) probabilities (see Details section for more information)."><meta property="og:image" content="https://docs.openrepgrid.org/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OpenRepGrid</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.18</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/web/intro.html">Intro</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/web/features.html">Features</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://eepurl.com/4h0t9" target="_blank" aria-label="Newsletter"><span class="fa fa-envelope fa-light"></span> Newsletter</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/markheckmann/OpenRepGrid/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Multiscale bootstrap cluster analysis.</h1>
      <small class="dont-index">Source: <a href="https://github.com/markheckmann/OpenRepGrid/blob/v0.1.18/R/calc.r" class="external-link"><code>R/calc.r</code></a></small>
      <div class="d-none name"><code>clusterBoot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>p-values are calculated for each branch of the cluster dendrogram to indicate the stability of a specific partition.
<code>clusterBoot</code> will yield the same clusters as the <code><a href="cluster.html">cluster()</a></code> function (i.e. standard hierarchical clustering) with
additional p-values. Two kinds of p-values are reported: bootstrap probabilities (BP) and approximately unbiased
(AU) probabilities (see Details section for more information).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">clusterBoot</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  along <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  align <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dmethod <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  cmethod <span class="op">=</span> <span class="st">"ward.D"</span>,</span>
<span>  p <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  nboot <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  r <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">1.4</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trim <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p><code>grid object</code></p></dd>


<dt id="arg-along">along<a class="anchor" aria-label="anchor" href="#arg-along"></a></dt>
<dd><p>Along which dimension to cluster. 1 = constructs, 2= elements.</p></dd>


<dt id="arg-align">align<a class="anchor" aria-label="anchor" href="#arg-align"></a></dt>
<dd><p>Whether the constructs should be aligned before clustering
(default is <code>TRUE</code>). If not, the grid matrix is clustered
as is. See Details section for more information.</p></dd>


<dt id="arg-dmethod">dmethod<a class="anchor" aria-label="anchor" href="#arg-dmethod"></a></dt>
<dd><p>The distance measure to be used. This must be one of "euclidean", "maximum", "manhattan", "canberra",
"binary" or "minkowski". Any unambiguous substring can be given. For additional information on the different types
type <code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">?dist</a></code>.</p></dd>


<dt id="arg-cmethod">cmethod<a class="anchor" aria-label="anchor" href="#arg-cmethod"></a></dt>
<dd><p>The agglomeration method to be used. This should be (an unambiguous abbreviation of) one of
<code>"ward.D"</code>, <code>"ward.D2"</code>, <code>"single"</code>, <code>"complete"</code>, <code>"average"</code>, <code>"mcquitty"</code>, <code>"median"</code> or <code>"centroid"</code>.</p></dd>


<dt id="arg-p">p<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>Power of the Minkowski metric. Not yet passed on to pvclust!</p></dd>


<dt id="arg-nboot">nboot<a class="anchor" aria-label="anchor" href="#arg-nboot"></a></dt>
<dd><p>the number of bootstrap replications. The default is
    <code>1000</code>.</p></dd>


<dt id="arg-r">r<a class="anchor" aria-label="anchor" href="#arg-r"></a></dt>
<dd><p>numeric vector which specifies the relative sample sizes of
    bootstrap replications. For original sample size \(n\) and
    bootstrap sample size \(n'\), this is defined as \(r=n'/n\).</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed for bootstrapping. Can be set for reproducibility (see
<code><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed()</a></code>). Usually not needed.</p></dd>


<dt id="arg-trim">trim<a class="anchor" aria-label="anchor" href="#arg-trim"></a></dt>
<dd><p>the number of characters a construct is trimmed to. If <code>NA</code> (default), no trimming is done.
Trimming simply saves space when displaying the output.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments to pass on to <code><a href="https://rdrr.io/pkg/pvclust/man/pvclust.html" class="external-link">pvclust::pvclust()</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A pvclust object as returned by the function <code><a href="https://rdrr.io/pkg/pvclust/man/pvclust.html" class="external-link">pvclust::pvclust()</a></code></p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>In standard (hierarchical) cluster analysis the question arises which of the identified structures are significant
or just emerged by chance. Over the last decade several methods have been developed to test structures for
robustness. One line of research in this area is based on resampling. The idea is to resample the rows or columns of
the data matrix and to build the dendrogram for each bootstrap sample (Felsenstein, 1985). The p-values indicates
the percentage of times a specific structure is identified across the bootstrap samples. It was shown that the
p-value is biased (Hillis &amp; Bull, 1993; Zharkikh &amp; Li, 1995). In the literature several methods for bias correction
have been proposed. In <code>clusterBoot</code> a method based on the
<em>multiscale bootstrap</em> is used to derive corrected (approximately
unbiased) p-values (Shimodaira, 2002, 2004). In conventional bootstrap analysis the size of the bootstrap sample is
identical to the original sample size. Multiscale bootstrap varies the bootstrap sample size in order to infer a
correction formula for the biased p-value on the basis of the variation of the results for the different sample
sizes (Suzuki &amp; Shimodaira, 2006).</p>
<p><strong>align</strong>: Aligning will reverse constructs if necessary to yield a
maximal similarity between constructs. In a first step the constructs are clustered including both directions. In a
second step the direction of a construct that yields smaller distances to the adjacent constructs is preserved and
used for the final clustering. As a result, every construct is included once but with an orientation that guarantees
optimal clustering. This approach is akin to the procedure used in FOCUS (Jankowicz &amp; Thomas, 1982).</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Felsenstein, J. (1985). Confidence Limits on Phylogenies: An Approach Using
the Bootstrap. <em>Evolution, 39</em>(4), 783. doi:10.2307/2408678</p>
<p>Hillis, D. M., &amp; Bull, J. J. (1993). An Empirical Test of Bootstrapping as a
Method for Assessing Confidence in Phylogenetic Analysis. <em>Systematic Biology,
42</em>(2), 182-192.</p>
<p>Jankowicz, D., &amp; Thomas, L. (1982). An Algorithm for the Cluster Analysis of
Repertory Grids in Human Resource Development. <em>Personnel Review,
11</em>(4), 15-22. doi:10.1108/eb055464.</p>
<p>Shimodaira, H. (2002) An approximately unbiased test of phylogenetic tree
selection. <em>Syst, Biol., 51</em>, 492-508.</p>
<p>Shimodaira,H. (2004) Approximately unbiased tests of regions using multistep-
multiscale bootstrap resampling. <em>Ann. Stat., 32</em>, 2616-2614.</p>
<p>Suzuki, R., &amp; Shimodaira, H. (2006). Pvclust: an R package for assessing the
uncertainty in hierarchical clustering. <em>Bioinformatics,
22</em>(12), 1540-1542. doi:10.1093/bioinformatics/btl117</p>
<p>Zharkikh, A., &amp; Li, W.-H. (1995). Estimation of confidence in phylogeny: the
complete-and-partial bootstrap technique. <em>Molecular Phylogenetic Evolution,
4</em>(1), 44-63.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># pvclust must be loaded</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://stat.sys.i.kyoto-u.ac.jp/prog/pvclust/" class="external-link">pvclust</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># p-values for construct dendrogram</span></span></span>
<span class="r-in"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu">clusterBoot</span><span class="op">(</span><span class="va">boeker</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/pvclust/man/pvpick.html" class="external-link">pvrect</a></span><span class="op">(</span><span class="va">s</span>, max.only <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># p-values for element dendrogram</span></span></span>
<span class="r-in"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu">clusterBoot</span><span class="op">(</span><span class="va">boeker</span>, along <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/pvclust/man/pvpick.html" class="external-link">pvrect</a></span><span class="op">(</span><span class="va">s</span>, max.only <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://markheckmann.de" class="external-link">Mark Heckmann</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Provided without <strong>any warranty</strong></p>
</div>

    </footer></div>





  </body></html>

